Correctly handle DNS redirects for NS servers that have no ADDITIONAL…

normanmaurer opened the pull request:
… record

Motiviation:

We incorrectly did ignore NS servers during redirect which had no ADDITIONAL record. This could at worse have the affect that we failed the query completely as none of the NS servers had a ADDITIONAL record. Beside this using a DnsCache to cache authoritative nameservers does not work in practise as we we need different features and semantics when cache these servers (for example we also want to cache unresolved nameservers and resolve these on the fly when needed).

Modifications:

- Correctly take NS records into account that have no matching ADDITIONAL record
- Correctly handle multiple ADDITIONAL records for the same NS record
- Introduce AuthoritativeDnsServerCache as a replacement of the DnsCache when caching authoritative nameservers + adding default implementation
- Add an adapter layer to reduce API breakage as much as possible
- Replace DnsNameResolver.uncachedRedirectDnsServerStream(...) with uncachedRedirectDnsServerList(...)
- Add unit tests

Result:

Our DnsResolver now correctly handle redirects in all cases.

normanmaurer commented:
@slandelle FYI as well...

trustin commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -465,13 +521,48 @@ private boolean handleRedirect(
+                }
+
+                if (!resolvedNameServers.isEmpty() || !unresolvedNameServers.isEmpty()) {
+                    System.err.println(resolvedNameServers);
```
!

trustin commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -774,6 +865,48 @@ private boolean query(String hostname, DnsRecordType type, DnsServerAddressStrea
+            }
+            InetSocketAddress address = originalStream.next();
+            if (address.equals(replaced)) {
+                resolved =  resolvedAddresses.iterator();
```
Double spaces

trustin commented on resolver-dns/src/test/java/io/netty/resolver/dns/DnsNameResolverTest.java:
```diff
@@ -1009,7 +1012,7 @@ public void testResolveAllHostsFile() {
                         if ("foo.com.".equals(inetHost)) {
                             try {
-                                return InetAddress.getByAddress("foo.com", new byte[] { 1, 2, 3, 4 });
+                                return InetAddress.getByAddress("foo.com", new byte[]{1, 2, 3, 4});
```
Missing space between `]` and `{`?

trustin commented on resolver-dns/src/test/java/io/netty/resolver/dns/DnsNameResolverTest.java:
```diff
@@ -1342,6 +1342,233 @@ int dnsRedirectPort(InetAddress server) {
+        };
+
+        try {
+            assertNotNull(resolver.resolveAll(hostname).await().cause());
```
Could we at least check exception type and message?

normanmaurer replied:
Sure...

trustin commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -338,51 +323,120 @@ private void query(final DnsServerAddressStream nameServerAddrStream,
+                        resolveCache(), new AuthoritativeDnsServerCache() {
+                    @Override
+                    public List<InetSocketAddress> get(String hostname) {
+                        // To not risk fall into any loop we will not use the cache during follow redirects but only
```
- fall -> falling
- comma before 'we'

trustin commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -338,51 +323,120 @@ private void query(final DnsServerAddressStream nameServerAddrStream,
-                    }
-                    return;
-                }
+        if (nameServerAddr.isUnresolved()) {
```
Could we extract the block guarded by this `if` into another method for readability?

trustin commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -338,51 +323,120 @@ private void query(final DnsServerAddressStream nameServerAddrStream,
+                }).resolve(resolverPromise);
             }
-        });
+        } else {
```
How about returning at the end of the `if` block above and unwrapping this `else` block?

trustin commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -465,13 +521,48 @@ private boolean handleRedirect(
+                    List<InetSocketAddress> addresses = new ArrayList<InetSocketAddress>(
+                            resolvedNameServers.size() + unresolvedNameServers.size());
+                    addresses.addAll(resolvedNameServers);
+                    addresses.addAll(unresolvedNameServers);
```
This logic that concats `resolvedNameServers` and `unresolvedNameServers` could be moved to the default implementation of `uncachedRedirectDnsServerList()`? By doing so, a custom `uncachedRedirectDnsServerList()` implementation may be able to construct the server list more efficiently.

normanmaurer replied:
Good idea

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DefaultAuthoritativeDnsServerCache.java:
```diff
@@ -0,0 +1,313 @@
+        if (maxTtl == 0) {
+            return;
+        }
+        if (PlatformDependent.javaVersion() >= 7 && address.getHostString() == null) {
```
No idea ... better safe then sorry ;)

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -153,6 +156,23 @@
+            if (!addr1.isUnresolved() && preferredAddressType().addressType() == addr1.getAddress().getClass()) {
+                return 1;
+            }
+            if (!addr2.isUnresolved() && preferredAddressType().addressType() == addr2.getAddress().getClass()) {
```
Maybe that is the safer bet... let me change it

normanmaurer commented:
@trustin can you please check again ?

carl-mastrangelo commented on resolver-dns/src/main/java/io/netty/resolver/dns/AuthoritativeDnsServerCache.java:
```diff
@@ -0,0 +1,64 @@
+     * {@link List} may contain unresolved {@link InetSocketAddress}es that will be resolved when needed while
+     * resolving other domain names.
+     *
+     * @param hostname the hostname
```
This should indicate if it returns `null` or an empty list if the host isn't found.

normanmaurer replied:
sounds good.

normanmaurer replied:
actually let me just change it to always return an empty List. This makes it also easier to handle.

normanmaurer replied:
done

carl-mastrangelo commented on resolver-dns/src/main/java/io/netty/resolver/dns/DefaultAuthoritativeDnsServerCache.java:
```diff
@@ -0,0 +1,319 @@
+        }
+
+        EntryList(List<Entry> entries) {
+            this.entries = entries;
```
Optional: make a defensive copy.

normanmaurer replied:
I think thats not needed with the current implementation.

carl-mastrangelo commented on resolver-dns/src/main/java/io/netty/resolver/dns/DefaultAuthoritativeDnsServerCache.java:
```diff
@@ -0,0 +1,319 @@
+ * Default implementation of {@link AuthoritativeDnsServerCache}, backed by a {@link ConcurrentMap}.
+ * */
+@UnstableApi
+public class DefaultAuthoritativeDnsServerCache implements AuthoritativeDnsServerCache {
```
Can I ask what happens to CNAME entries?  Do they get cached by their original name, or destination name?

normanmaurer replied:
should be the original name.

normanmaurer replied:
@carl-mastrangelo let me add a test-case ...

carl-mastrangelo commented on resolver-dns/src/main/java/io/netty/resolver/dns/DefaultAuthoritativeDnsServerCache.java:
```diff
@@ -0,0 +1,319 @@
+    }
+
+    @Override
+    public List<InetSocketAddress> get(String hostname) {
```
Will this ever support other record types (like SRV and TXT), or just A / AAAA?

normanmaurer replied:
@carl-mastrangelo we only care about A / AAAA here as this is what we can use as a name server.

carl-mastrangelo commented on resolver-dns/src/main/java/io/netty/resolver/dns/AuthoritativeDnsServerCache.java:
```diff
@@ -0,0 +1,64 @@
+     * @param originalTtl the TTL as returned by the DNS server
+     * @param loop the {@link EventLoop} used to register the TTL timeout
+     */
+    void cache(String hostname, InetSocketAddress address, long originalTtl, EventLoop loop);
```
This API is odd, because DNS always gives you back lots of addresses at the same TTL, not one.  If you had to call this in a loop, the TTLs might slide.  This would be surprising as a user.  

Consider taking a list of records for a given TTL.

normanmaurer replied:
@carl-mastrangelo while the TTL may be the same its not required...

```
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 62895
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 4, ADDITIONAL: 9
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;ns1.google.com.			IN	A

;; AUTHORITY SECTION:
google.com.		172800	IN	NS	ns2.google.com.
google.com.		172800	IN	NS	ns1.google.com.
google.com.		172800	IN	NS	ns3.google.com.
google.com.		172800	IN	NS	ns4.google.com.

;; ADDITIONAL SECTION:
ns2.google.com.		172800	IN	AAAA	2001:4860:4802:34::a
ns2.google.com.		172800	IN	A	216.239.34.10
ns1.google.com.		172800	IN	AAAA	2001:4860:4802:32::a
ns1.google.com.		172800	IN	A	216.239.32.10
ns3.google.com.		172800	IN	AAAA	2001:4860:4802:36::a
ns3.google.com.		172800	IN	A	216.239.36.10
ns4.google.com.		172800	IN	AAAA	2001:4860:4802:38::a
ns4.google.com.		172800	IN	A	216.239.38.10
```

As you can see each of these define their own TTL. Sure its the same here but this is not required by the protocol so we should not make assumptions here.

carl-mastrangelo replied:
In the case the TTL is the same, I am suggesting that they be added atomically.   If the cache is queried right as the scheduled removal is happening, the cache may return partial results.  This is what I meant by sliding.

Also, not clear on the expected performance needs here, but I would also assume a single scheduled future is lighter weight than four.

trustin replied:
+1 for consolidating cache records but perhaps we could take care of it in a follow-up PR to fix both `DnsCache` and `AuthoritativeDnsServerCache`?

normanmaurer replied:
At least in our current implementation it’s not a problem as we just remove all entries once the smallest TTL matches. This may result in more queries then needed but is the safer bet to not end up with partial results imho:

https://github.com/netty/netty/pull/8177/files#diff-7ab1cd3db010ccc37c01987e522675caR178

That said I guess we could also just schedule for the smallest TTL here. 

WDYT?

normanmaurer replied:
Here is a PR for the "smallest TTL change" for the `DefaultDnsCache`:

https://github.com/netty/netty/pull/8187/

Something like this could also be used for the `DefaultAuthoritativeDnsServerCache` here. Let me know what you think @carl-mastrangelo @trustin

trustin replied:
Sounds good to me.

carl-mastrangelo commented on resolver-dns/src/main/java/io/netty/resolver/dns/DefaultAuthoritativeDnsServerCache.java:
```diff
@@ -0,0 +1,319 @@
+            super(new EntryList(entry));
+        }
+
+        Entry add(Entry e) {
```
Why not just use CopyOnWriteArrayList ?

normanmaurer replied:
@carl-mastrangelo let me think about it... I just followed what we have in `DefaultDnsCache`.

normanmaurer replied:
Actual I think the current approach is better... If we find out that `CopyOnWriteArrayList` is better we can still do a followup. For example you could do a PR if you think it worth it... WDYT @carl-mastrangelo

carl-mastrangelo replied:
It doesn't seem like a big issue.   You can leave it as is.

normanmaurer commented:
Also @vietj

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -153,6 +156,25 @@
+                    preferredAddressType().addressType().isAssignableFrom(addr2.getAddress().getClass())) {
+                return -1;
+            }
+            return 1;
```
IIUC this `Comparator` violates the interface:

```java
     * The implementor must ensure that <tt>sgn(compare(x, y)) ==
     * -sgn(compare(y, x))</tt> for all <tt>x</tt> and <tt>y</tt>.  (This
     * implies that <tt>compare(x, y)</tt> must throw an exception if and only
     * if <tt>compare(y, x)</tt> throws an exception.)<p>
```

example:
```java
InetSocketAddress addr1 = InetSocketAddress.createUnresolved("foo", 123);
        InetSocketAddress addr2 = InetSocketAddress.createUnresolved("bar", 456);
        int x = nameServerComparator.compare(addr1, addr2);
        int y = nameServerComparator.compare(addr2, addr1);
        assert x == -y;
```

there is also the transitive property to consider:

```
     * The implementor must also ensure that the relation is transitive:
     * <tt>((compare(x, y)&gt;0) &amp;&amp; (compare(y, z)&gt;0))</tt> implies
     * <tt>compare(x, z)&gt;0</tt>.<p>
```

Also if there are multiple addresses that are preferred & resolved the comparator may have issues.

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -440,20 +505,22 @@ private boolean handleRedirect(
 
             if (serverNames != null) {
-                List<InetSocketAddress> nameServers = new ArrayList<InetSocketAddress>(serverNames.size());
+                Set<InetSocketAddress> resolvedNameServers = parent.newNameServerSet();
```
using a set may re-order the results provided by the server, which may be in preference order from the server's perspective. Should we just depend upon the `uncachedRedirectDnsServerList` below to apply preference/sorting order so the user can see the original order and respect it if desired?

normanmaurer replied:
@Scottmitch good point.

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -465,13 +532,40 @@ private boolean handleRedirect(
+                    // Process all unresolved nameservers as well.
+                    for (;;) {
+                        final AuthoritativeNameServer authoritativeNameServer = serverNames.handleNext();
+                        if (authoritativeNameServer == null) {
```
nit: you could use a do/while loop here because you already checked there is at least one element.

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -368,21 +359,95 @@ public void operationComplete(Future<AddressedEnvelope<DnsResponse, InetSocketAd
+                public List<InetSocketAddress> get(String hostname) {
+                    // To not risk falling into any loop, we will not use the cache during follow redirects but only
+                    // on the initial query.
+                    return Collections.emptyList();
```
Discussed with @normanmaurer offline ... fine as is.
- A scenario where caching the unresolved address is if there is a discrepancy in TTLs between the first name, and the second (or n-th) name. If the unresolved address has relatively long TTL and is slower to resolve, and second (or n-th) has a shorter TTL and resolves more quickly then there is value in caching the unresolved address.
- Also caches can opt out of storing unresolved addresses if they want to.

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DefaultAuthoritativeDnsServerCache.java:
```diff
@@ -0,0 +1,319 @@
+                            // Replace the old entry.
+                            assert replacedEntry == null;
+                            replacedEntry = entry;
+                            newEntries.add(e);
```
I think I would like to keep this as it will trigger the assert then.

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DefaultAuthoritativeDnsServerCache.java:
```diff
@@ -0,0 +1,319 @@
+            ttl = originalTtl;
+        }
+        scheduleCacheExpiration(key, e, Math.max(minTtl,
+                Math.min(MAX_SUPPORTED_TTL_SECS, (int) Math.min(maxTtl, ttl))), loop);
```
nope... good catch.

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -153,6 +156,25 @@
+            if (addr1.equals(addr2)) {
+                return 0;
+            }
+            if (!addr1.isUnresolved() &&
```
Good point... I used it before also for unresolved addresses. We need it to ensure that if we have AAAA and A  addresses we use the "preferred type" first.

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -280,44 +294,15 @@ private DnsServerAddressStream getNameServersFromCache(String hostname) {
-            List<? extends DnsCacheEntry> entries = parent.authoritativeDnsServerCache().get(hostname, additionals);
-            if (entries != null && !entries.isEmpty()) {
-                return DnsServerAddresses.sequential(new DnsCacheIterable(entries)).stream();
+            List<InetSocketAddress> entries = authoritativeDnsServerCache().get(hostname);
```
I am not sure about this.. I mean we could do this but its also not too bad to return a `List` here imho. At least in our `DefaultAuthoritativeDnsServerCache` we do not need to do an allocation for this ;) 

@trustin @carl-mastrangelo thoughts ?

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DefaultAuthoritativeDnsServerCache.java:
```diff
@@ -0,0 +1,319 @@
+
+        final Entry e = new Entry(address);
+
+        String key = appendDot(hostname);
```
Like discussed offline I think you are right and we should not append anything here... Let me change this.

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DefaultAuthoritativeDnsServerCache.java:
```diff
@@ -0,0 +1,319 @@
+        if (old != null) {
+            ScheduledFuture<?> future = old.expirationFuture;
+            if (future != null) {
+                ttl = Math.min(originalTtl, future.getDelay(TimeUnit.SECONDS));
```
I think we should just port over the changes that we did in https://github.com/netty/netty/pull/8187 once it is merged. This will also "resolve" the comment.

normanmaurer replied:
@Scottmitch done via baa3fd31638215828cb80af457eed1974a2fd209. If we adjust  #8187 I will bring over the changes.

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -368,21 +359,95 @@ public void operationComplete(Future<AddressedEnvelope<DnsResponse, InetSocketAd
+        assert nameServerName != null;
+
+        if (nameServerName.equals(question.name())) {
+            // If the nameserver is the same as what we want to lookup we should skip it.
```
Actually I think it should never... Removed the code

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -324,11 +400,20 @@ final DnsQueryLifecycleObserverFactory dnsQueryLifecycleObserverFactory() {
      */
-    protected DnsServerAddressStream uncachedRedirectDnsServerStream(List<InetSocketAddress> nameServers) {
-        return DnsServerAddresses.sequential(nameServers).stream();
+    protected List<InetSocketAddress> uncachedRedirectDnsServerList(
```
The problem is then we need to open up a way for the user to create an `DnsServerAddressStream` which contains unresolved addresses. We are currently not allowing this by the static methods that are exposed by `DnsServerAddresses`. Not saying we can not change this... I just wonder if we should or not... Thoughts ?

trustin replied:
Probably OK to allow unresolved addresses, as long as documented properly? For example, we could:

- add a method like `DnsServerAddressStream.containsUnresolved()`.
- the implementations returned by the factory methods in `DnsServerAddresses` could implement it correctly.
- we could fail or resolve (and cache) the unresolved addresses with system-configured name servers if it returns false.

normanmaurer commented:
@Scottmitch I think once this and #8187 is merged we can reduce a bit of code duplication in the caches (related to 
https://github.com/netty/netty/pull/8177/commits/baa3fd31638215828cb80af457eed1974a2fd209) by creating something that can be used by both internally.

normanmaurer commented:
@carl-mastrangelo can you re-check as well ?

trustin commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -401,19 +385,14 @@ final DnsQueryLifecycleObserverFactory dnsQueryLifecycleObserverFactory() {
-     * @param resolvedNameServers The resolved addresses of the DNS servers which are used in the event of a redirect.
-     * @param unresolvedNameServers The unresolved addresses of the DNS servers which are used in the event of
-     *                              a redirect.
+     * @param nameservers The  addresses of the DNS servers which are used in the event of a redirect. This may
```
nit: double spaces

normanmaurer commented:
@Scottmitch I replaced `List` usage by `DnsServerAddressStream` and addressed also your other comments. Please check again. Also maybe @trustin want to check as well again.

Please note I did not expose our `DnsSterverAddressStream` implementation to the user that allows unresolved `InetSocketAddresses` for now as I suspect the user want to write their own implementation anyway if a custom cache is implemented or `uncachedRedirectDnsServerStream` is overridden.

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -467,20 +526,97 @@ private boolean handleRedirect(
+                    if (authoritativeNameServer == null) {
+                        break;
+                    }
+                    // These will be resolved on the fly if needed.
```
what if these entries exist in the cache already? The default comparator in `uncachedRedirectDnsServerStream` may prefer resolved addresses when compared to unresolved addresses. However it is possible that we are altering the order in which results are returned in the `RESPONSE` record because the server didn't directly provide the IP address, but we may already know the resolved address via the cache. I think we should cache this intermediate cname/redirect element, but also try to satisfy the IP address(es) via the cache.

normanmaurer replied:
@Scottmitch I think I dont understand your comment ... Can you explain more what problems you see ? Is it because the order in the cache may be different then the order that is used by `uncachedRedirectDnsServerStream` ? Are you saying we should use the same order and even only put in the cache what is returned by  `uncachedRedirectDnsServerStream `?

normanmaurer replied:
@Scottmitch @trustin @carl-mastrangelo WDYT about the "ordering" question here ?

Scottmitch replied:
the comment was meant to highlight that we get results back from the server (e.g. Authority: A, B, C; Additional: B:IPx) and by default we reorder them according to which ones have the IPs specified in the additional section (B:IPx, A, C) ... however even if the server doesn't include a corresponding Additional entry, we may know the IP from the cache and can preserve the original ordering from the server (e.g. A:IPy, B:IPx, C)

IIRC the server responds in precedence order, so we should make a best effort to preserve this. We may want to prioritize name servers that we have a resolved address for (from cache or additional section), but potentially the unresolved DNS server will be more localized for the user and end up providing better performance in the long run?

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -467,20 +526,97 @@ private boolean handleRedirect(
+                            question.name(), nameServers);
+                    if (serverList != null) {
+                        query(serverList, 0, question,
+                              queryLifecycleObserver.queryRedirected(new DnsAddressStreamList(serverList)),
```
future discussion: here is another discrepancy between list/stream like APIs that we should consider making consistent later.

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -830,33 +1008,57 @@ void add(DnsRecord r) {
+            AuthoritativeNameServer serverName = head;
+
+            while (serverName != null) {
+                if (serverName.nsName.equalsIgnoreCase(nsName)) {
```
should this only "handle" the node if it has not yet been handled? For example what about and `ADDITIONAL` section where the same name server name is mapped to multiple IP addresses types.

```
;; ADDITIONAL SECTION:
ns2.google.com.		172800	IN	AAAA	2001:4860:4802:34::a
ns2.google.com.		172800	IN	A	216.239.34.10
```

Do we have a test for this?

```java
if (!serverName.handled && serverName.nsName.equalsIgnoreCase(nsName)) {
..
}
```

normanmaurer replied:
No we should actually handle both as otherwise one address will be missing in the result. This was a bug we had before (there is a test for this as well now).

Scottmitch replied:
I agree that we should handle both. This code looks like it will just stop at the first element with a matching name regardless if it is handled, and the subsequent element will never be marked "handled". So it seems like the state of this collection won't be correct in the presence of multiple entries with the same name.

normanmaurer replied:
ah got it... So you say we should always walk the whole list as there may be more then one that matches ?

Scottmitch replied:
I misunderstood how this was being used ... so disregard the original suggestion. However it is possible there will be duplicate NS (with different domain) in the authority section that would be problematic (not sure if this is possible).

normanmaurer replied:
I think this could not happen. So it should be fine ?

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsServerAddresses.java:
```diff
@@ -106,15 +104,15 @@ public static DnsServerAddresses shuffled(InetSocketAddress... addresses) {
             @Override
             public DnsServerAddressStream stream() {
-                return new ShuffledDnsServerAddressStream(addresses);
+                return new ShuffledDnsServerAddressStream(addresses.toArray(new InetSocketAddress[0]));
```
In the interest of avoiding conversions, if we are going to change from array to `List` can we either:
- make this change consistent across stream types (e.g. also to `ShuffledDnsServerAddressStream`)
- do the conversion once (instead of on each call to `stream()`)

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/NameServerComparator.java:
```diff
@@ -0,0 +1,72 @@
+                    return 0;
+                }
+                return -1;
+            } else {
```
nit: consider refactoring for less indentation:

```diff
diff --git a/resolver-dns/src/main/java/io/netty/resolver/dns/NameServerComparator.java b/resolver-dns/src/main/java/io/netty/resolver/dns/NameServerComparator.java
index b2c8d7a0f2..7cced6ddd3 100644
--- a/resolver-dns/src/main/java/io/netty/resolver/dns/NameServerComparator.java
+++ b/resolver-dns/src/main/java/io/netty/resolver/dns/NameServerComparator.java
@@ -57,13 +57,11 @@ final class NameServerComparator implements Comparator<InetSocketAddress> {
                     return 0;
                 }
                 return -1;
-            } else {
-                if (addr2.isUnresolved()) {
-                    // Prefer the resolved address.
-                    return -1;
-                }
-                return 1;
+            } else if (addr2.isUnresolved()) {
+                // Prefer the resolved address.
+                return -1;
             }
+            return 1;
         } else if (!addr2.isUnresolved()) {
             return 1;
         }
```

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/NameServerComparator.java:
```diff
@@ -0,0 +1,72 @@
+
+    @Override
+    public int compare(InetSocketAddress addr1, InetSocketAddress addr2) {
+        if (addr1 == addr2) {
```
should this be `equals` instead of reference comparison?

normanmaurer replied:
nope... as we use we want to not use the default `equals` implementation. Let me add a comment

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/NameServerComparator.java:
```diff
@@ -0,0 +1,72 @@
+            return 0;
+        }
+        if (!addr1.isUnresolved()) {
+            if (!addr2.isUnresolved()) {
```
consider refactoring to reduce embedded conditionals:

```diff
+diff --git a/resolver-dns/src/main/java/io/netty/resolver/dns/NameServerComparator.java b/resolver-dns/src/main/java/io/netty/resolver/dns/NameServerComparator.java
index b2c8d7a0f2..6b179d85db 100644
--- a/resolver-dns/src/main/java/io/netty/resolver/dns/NameServerComparator.java
+++ b/resolver-dns/src/main/java/io/netty/resolver/dns/NameServerComparator.java
@@ -43,12 +43,10 @@ final class NameServerComparator implements Comparator<InetSocketAddress> {
             return 0;
         }
         if (!addr1.isUnresolved()) {
-            if (!addr2.isUnresolved()) {
-                if (addr1.getAddress().getClass() == addr2.getAddress().getClass()) {
-                    // If both addresses use the same type we will just return 0 to preserve the original order
-                    // when sorting the List.
-                    return 0;
-                }
+            if (!addr2.isUnresolved() && addr1.getAddress().getClass() == addr2.getAddress().getClass()) {
+                // If both addresses use the same type we will just return 0 to preserve the original order
+                // when sorting the List.
+                return 0;
             }
             if (preferredAddressType.isAssignableFrom(addr1.getAddress().getClass())) {
                 if (!addr2.isUnresolved() && preferredAddressType.isAssignableFrom(addr2.getAddress().getClass())) {

```

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/NameServerComparator.java:
```diff
@@ -0,0 +1,72 @@
+        if (addr1 == addr2) {
+            return 0;
+        }
+        if (!addr1.isUnresolved()) {
```
consider using the following more compact comparison which is more compact and easier to follow IMHO:

```java
        // we prioritize resolved addresses that are of the preferred type
        if (addr1.equals(addr2)) {
            return 0;
        } else if (!addr1.isUnresolved() && !addr2.isUnresolved()) {
            if (addr1.getAddress().getClass() == addr2.getAddress().getClass()) {
                return 0;
            }
            return preferredAddressType.isAssignableFrom(addr1.getAddress().getClass()) ? -1 : 1;
        } else if (addr1.isUnresolved()) {
            return 1;
        } else if (addr2.isUnresolved()) {
            return -1;
        }
        return 0;
```

normanmaurer replied:
thats not the same an even fails a unit test...

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DefaultAuthoritativeDnsServerCache.java:
```diff
@@ -0,0 +1,308 @@
+     *
+     * @see #maxTtl()
+     */
+    public int minTtl() {
```
No I guess we can just remove these for now.

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/AuthoritativeDnsServerCacheAdapter.java:
```diff
@@ -0,0 +1,80 @@
+        do {
+            InetAddress addr = entries.get(i).address();
+            addresses.add(new InetSocketAddress(addr, DefaultDnsServerAddressStreamProvider.DNS_PORT));
+        } while (++i < entries.size());
```
@Scottmitch yeah I didn't want to depend on the semantics of `DefaultDnsCache` as its an imp detail imho. Also this is just a "translation" layer to keep API breakage to a minimum for now which will be removed later so I would not worry about improving this a lot.

normanmaurer replied:
maybe... not sure. That said I thought better safe then sorry. And again its really just something to give people time to upgrade to the new API, so I am not really worried by some extra copies. If you are super strong on this I can change it tho.

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DefaultAuthoritativeDnsServerCache.java:
```diff
@@ -0,0 +1,308 @@
+                    int i = 0;
+                    do {
+                        InetSocketAddress address = entryList.get(i);
+                        if (replacedEntry != null || !address.getHostName().equalsIgnoreCase(nameServerName)) {
```
sure why not

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/NameServerComparator.java:
```diff
@@ -0,0 +1,72 @@
+
+    @Override
+    public int compare(InetSocketAddress addr1, InetSocketAddress addr2) {
+        // We dont want to use equals(...) here as we dont want to use the default implementation.
```
`dont` -> `don't`

can you explain *why* we don't want to use the default implementation?

normanmaurer replied:
see 
https://github.com/netty/netty/pull/8177#discussion_r210184929

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/NameServerComparator.java:
```diff
@@ -0,0 +1,72 @@
+            return 0;
+        }
+
+        if (!addr1.isUnresolved()) {
```
continuation of https://github.com/netty/netty/pull/8177#discussion_r21004160 ... fair point the previous suggestion failed unit tests. The point I was trying to make is the conditional logic here seems a bit convoluted and there is an opportunity to simplify. Here is a reworked suggestion that passed unit tests.

```java
        // we prioritize resolved addresses that are of the preferred type
        if (addr1.equals(addr2)) { // maybe use == instead?
            return 0;
        } else if (!addr1.isUnresolved() && !addr2.isUnresolved()) {
            if (addr1.getAddress().getClass() == addr2.getAddress().getClass()) {
                return 0;
            }
            return preferredAddressType.isAssignableFrom(addr1.getAddress().getClass()) ? -1 : 1;
        } else if (addr1.isUnresolved() && addr2.isUnresolved()) {
            return 0;
        }
        return addr1.isUnresolved() ? 1 : -1;
```

normanmaurer replied:
Good idea... I did apply this with some small changes as most `else` are not needed. Also after thinking more about it we can just use `equals` as you suggested. Sorry for the noise

normanmaurer commented:
@carl-mastrangelo sounds good ... :) I mean at worse I can do a follow up. I know its quite big but there was so much stuff broken :(

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -830,49 +972,166 @@ void add(DnsRecord r) {
+        /**
+         * Add an authoritative nameserver to the cache if its not a root server.
+         */
+        void cache(AuthoritativeDnsServerCache cache, InetSocketAddress address, EventLoop loop) {
```
I think we may have cross some wires in our offline discussion ... I think there are 2 parts to this puzzle.

#### Part 1 (cache insertion order)
my assumption was that we either cache insertion order is authority order, or we make a "best effort" to preserve server authority order if it avoids additional iteration/computation/memory. I don't think we want/need to preserve the order which the `uncachedRedirectDnsServerStream` method defines. If the user wants a custom order of elements in the cache, they can provide a custom cache implementation.

lets assume the following dns response:
```
AUTHORITY
foo.com NS ns1.foo.com.
foo.com NS ns2.foo.com.
foo.com NS ns3.foo.com.
foo.com NS ns4.foo.com.

ADDITIONAL
 ns2.foo.com. IPa
 ns4.foo.com. IPb
```

either the order we put into the cache is (server authority order):

```
ns1.foo.com. <unresovled, assuming not already in the cache>
ns2.foo.com IPa
ns3.foo.com. <unresovled, assuming not already in the cache>
ns4.foo.com IPb
```

or (best effort server authority order, preference for entries with additionals):

```
ns2.foo.com IPa
ns4.foo.com IPb
ns1.foo.com. <unresovled, assuming not already in the cache>
ns3.foo.com. <unresovled, assuming not already in the cache>
```

The `assuming not already in the cache` part is because we look up entries in the local cache which do not have any any additional entry to provide an IP (to try to cut down on DNS queries if we already know the answer). In `handleAll` `Long.MIN_VALUE` is currently used as a special value to treat differently in this case, but I don't think we need/want to modify the cache at all in this case.
The second option lines up with our algorithm/data-structures a bit more, so it maybe preferable.

#### Part 2 (ordering of redirect servers to query):
We can apply preference ordering to the redirect results be default, but can we clarify the `uncachedRedirectDnsServerStream` method name? This method name is confusing because we are actually caching results and entries maybe cached within. For example `newRedirectDnsServerStream` (with appropriate javadocs) would help avoid this confusion.

normanmaurer replied:
ok got it... this will also make things easier :)

normanmaurer replied:
For 1) I think we want to preserve "server authority order". 
 
For 2) renaming the method sounds good.

normanmaurer replied:
done ... I wonder why IntelliJ did not highlight it :(

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -830,49 +972,166 @@ void add(DnsRecord r) {
+                    InetAddress resolved = decodeAddress(r, nsName, parent.isDecodeIdn());
+                    if (resolved == null) {
+                        // Could not parse it, move to the next.
+                        continue;
```
Good catch... I just move the parsing also out of the loop as we only need to do it once.

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -830,49 +972,166 @@ void add(DnsRecord r) {
+                        serverName.next = server;
+                        serverName = server;
+
+                        count++;
```
sure

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -830,49 +972,166 @@ void add(DnsRecord r) {
+                        }
+                        AuthoritativeNameServer server = new AuthoritativeNameServer(serverName);
+                        server.next = serverName.next;
+                        serverName.next = server;
```
I added a test-case that show prove the current implementation already does what you describe and so works as expected.

normanmaurer replied:
yep.. I think it is fine as it is is.

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -830,49 +972,166 @@ void add(DnsRecord r) {
+        /**
+         * Creates a new {@link List} which holds the {@link InetSocketAddress}es.
+         */
+        List<InetSocketAddress> addressList() {
```
I think I made this better with my latest changes (I removed one iteration).

Scottmitch replied:
Here is what I had in mind to avoid this additional iteration at the cost of potentially copy/resize of the resulting list. It depends upon the use case but *maybe* resizing vs *always* doing another iteration maybe a good tradeoff to start. We didn't even handle the "no-additionals" case before this, so not sure how common this is?

```diff
diff --git a/resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java b/resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java
index b4c83e4c67..069c97dccf 100644
--- a/resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java
+++ b/resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java
@@ -487,6 +487,12 @@ abstract class DnsResolveContext<T> {
         if (res.count(DnsSection.ANSWER) == 0) {
             AuthoritativeNameServerList serverNames = extractAuthoritativeNameServers(question.name(), res);
             if (serverNames != null) {
+                // Make a best guess for how many elements there will be in the redirect list. This amount is variable
+                // because authority entries with no corresponding additional entries may result in adding a variable
+                // amount of addresses to this list from the cache.
+                List<InetSocketAddress> redirectedNameServerAddresses =
+                        new ArrayList<InetSocketAddress>(serverNames.getNameServerCount());
+                AuthoritativeDnsServerCache authoritativeDnsServerCache = authoritativeDnsServerCache();
                 int additionalCount = res.count(DnsSection.ADDITIONAL);
 
                 for (int i = 0; i < additionalCount; i++) {
@@ -499,17 +505,17 @@ abstract class DnsResolveContext<T> {
 
                     // We may have multiple ADDITIONAL entries for the same nameserver name. For example one AAAA and
                     // one A record.
-                    serverNames.handleWithAdditional(parent, r, authoritativeDnsServerCache());
+                    serverNames.handleWithAdditional(parent, r, authoritativeDnsServerCache,
+                            redirectedNameServerAddresses);
                 }
 
                 // Process all unresolved nameservers as well.
-                serverNames.handleWithoutAdditionals(parent, resolveCache(), authoritativeDnsServerCache());
-
-                List<InetSocketAddress> addresses = serverNames.addressList();
+                serverNames.handleWithoutAdditionals(parent, resolveCache(), authoritativeDnsServerCache,
+                        redirectedNameServerAddresses);
 
                 // Give the user the chance to sort or filter the used servers for the query.
                 DnsServerAddressStream serverStream = parent.newRedirectDnsServerStream(
-                        question.name(), addresses);
+                        question.name(), redirectedNameServerAddresses);
 
                 if (serverStream != null) {
                     query(serverStream, 0, question,
@@ -920,10 +926,10 @@ abstract class DnsResolveContext<T> {
 
         private final String questionName;
 
-        // We not expect the linked-list to be very long so a double-linked-list is overkill.
+        // We do not expect the linked-list to be very long so a double-linked-list is overkill.
         private AuthoritativeNameServer head;
 
-        private int servers;
+        private int nameServerCount;
 
         AuthoritativeNameServerList(String questionName) {
             this.questionName = questionName.toLowerCase(Locale.US);
@@ -967,20 +973,29 @@ abstract class DnsResolveContext<T> {
             // We are only interested in preserving the nameservers which are the closest to our qName, so ensure
             // we drop servers that have a smaller dots count.
             if (head == null || head.dots < dots) {
-                servers = 1;
                 head = new AuthoritativeNameServer(dots, r.timeToLive(), recordName, domainName);
+                nameServerCount = 1;
             } else if (head.dots == dots) {
                 AuthoritativeNameServer serverName = head;
                 while (serverName.next != null) {
                     serverName = serverName.next;
                 }
                 serverName.next = new AuthoritativeNameServer(dots, r.timeToLive(), recordName, domainName);
-                servers++;
+                ++nameServerCount;
             }
         }
 
+        boolean isEmpty() {
+            return head == null;
+        }
+
+        int getNameServerCount() {
+            return nameServerCount;
+        }
+
         void handleWithAdditional(
-                DnsNameResolver parent, DnsRecord r, AuthoritativeDnsServerCache authoritativeCache) {
+                DnsNameResolver parent, DnsRecord r, AuthoritativeDnsServerCache authoritativeCache,
+                List<InetSocketAddress> redirectedNameServerAddresses) {
             // Just walk the linked-list and mark the entry as handled when matched.
             AuthoritativeNameServer serverName = head;
 
@@ -1003,8 +1018,7 @@ abstract class DnsResolveContext<T> {
                         server.next = serverName.next;
                         serverName.next = server;
                         serverName = server;
-
-                        servers++;
+                        ++nameServerCount; // optional, this isn't used after this point
                     }
                     // We should replace the TTL if needed with the one of the ADDITIONAL record so we use
                     // the smallest for caching.
@@ -1012,6 +1026,7 @@ abstract class DnsResolveContext<T> {
 
                     // Cache the server now.
                     cache(serverName, authoritativeCache, parent.executor());
+                    redirectedNameServerAddresses.add(serverName.address);
                     return;
                 }
                 serverName = serverName.next;
@@ -1020,17 +1035,16 @@ abstract class DnsResolveContext<T> {
 
         // Now handle all AuthoritativeNameServer for which we had no ADDITIONAL record
         void handleWithoutAdditionals(
-                DnsNameResolver parent, DnsCache cache, AuthoritativeDnsServerCache authoritativeCache) {
+                DnsNameResolver parent, DnsCache cache, AuthoritativeDnsServerCache authoritativeCache,
+                List<InetSocketAddress> redirectedNameServerAddresses) {
             AuthoritativeNameServer serverName = head;
 
             while (serverName != null) {
                 if (serverName.address == null) {
                     // These will be resolved on the fly if needed.
                     cacheUnresolved(serverName, authoritativeCache, parent.executor());
-                    servers++;
 
                     // Try to resolve via cache as we had no ADDITIONAL entry for the server.
-
                     List<? extends DnsCacheEntry> entries = cache.get(serverName.nsName, null);
                     if (entries != null && !entries.isEmpty()) {
                         InetAddress address = entries.get(0).address();
@@ -1038,6 +1052,7 @@ abstract class DnsResolveContext<T> {
                         // If address is null we have a resolution failure cached so just use an unresolved address.
                         if (address != null) {
                             serverName.update(parent.newRedirectServerAddress(address));
+                            redirectedNameServerAddresses.add(serverName.address);
 
                             for (int i = 1; i < entries.size(); i++) {
                                 address = entries.get(i).address();
@@ -1049,10 +1064,14 @@ abstract class DnsResolveContext<T> {
                                 serverName.next = server;
                                 serverName = server;
                                 serverName.update(parent.newRedirectServerAddress(address));
-
-                                servers++;
+                                ++nameServerCount; // optional, this isn't used after this point
+                                redirectedNameServerAddresses.add(serverName.address);
                             }
+                        } else {
+                            redirectedNameServerAddresses.add(serverName.address);
                         }
+                    } else {
+                        redirectedNameServerAddresses.add(serverName.address);
                     }
                 }
                 serverName = serverName.next;
@@ -1061,10 +1080,6 @@ abstract class DnsResolveContext<T> {
 
         private static void cacheUnresolved(
                 AuthoritativeNameServer server, AuthoritativeDnsServerCache authoritativeCache, EventLoop loop) {
-            // As we have no idea about how long the TTL should be we will just cache the unresolved
-            // address
-
-            // We still want to cached the unresolved address
             server.address = InetSocketAddress.createUnresolved(
                     server.nsName, DefaultDnsServerAddressStreamProvider.DNS_PORT);
 
@@ -1074,40 +1089,9 @@ abstract class DnsResolveContext<T> {
 
         private static void cache(AuthoritativeNameServer server, AuthoritativeDnsServerCache cache, EventLoop loop) {
             // Cache NS record if not for a root server as we should never cache for root servers.
-            if (!server.isRootServer() && server.address != null) {
-                // If we resolved the AuthoritativeNameServer via the DnsCache before we should not cache the
-                // resolved address as we don't have a good idea about the TTL to apply. We will just cache
-                // the unresolved address for now. We will then query the cache again the next time we try to
-                // use it for a query which will do the correct thing in terms of respecting the TTL of the
-                // A / AAAA record.
-                InetSocketAddress addressToCache = server.ttl == Long.MIN_VALUE ?
-                        InetSocketAddress.createUnresolved(server.nsName, server.address.getPort()) :
-                        server.address;
-                cache.cache(server.domainName, addressToCache, server.ttl, loop);
-            }
-        }
-
-        /**
-         * Returns {@code true} if empty, {@code false} otherwise.
-         */
-        boolean isEmpty() {
-            return servers == 0;
-        }
-
-        /**
-         * Creates a new {@link List} which holds the {@link InetSocketAddress}es.
-         */
-        List<InetSocketAddress> addressList() {
-            List<InetSocketAddress> addressList = new ArrayList<InetSocketAddress>(servers);
-
-            AuthoritativeNameServer server = head;
-            while (server != null) {
-                if (server.address != null) {
-                    addressList.add(server.address);
-                }
-                server = server.next;
+            if (!server.isRootServer()) {
+                cache.cache(server.domainName, server.address, server.ttl, loop);
             }
-            return addressList;
         }
     }
 ```

normanmaurer replied:
So after talking with Scott we will keep it as it is as it often happens that the NS record count is not a good way to size the list.

For example it may see:

```
; <<>> DiG 9.10.6 <<>> @192.12.94.30 ns1.google.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 13399
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 4, ADDITIONAL: 9
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;ns1.google.com.			IN	A

;; AUTHORITY SECTION:
google.com.		172800	IN	NS	ns2.google.com.
google.com.		172800	IN	NS	ns1.google.com.
google.com.		172800	IN	NS	ns3.google.com.
google.com.		172800	IN	NS	ns4.google.com.

;; ADDITIONAL SECTION:
ns2.google.com.		172800	IN	AAAA	2001:4860:4802:34::a
ns2.google.com.		172800	IN	A	216.239.34.10
ns1.google.com.		172800	IN	AAAA	2001:4860:4802:32::a
ns1.google.com.		172800	IN	A	216.239.32.10
ns3.google.com.		172800	IN	AAAA	2001:4860:4802:36::a
ns3.google.com.		172800	IN	A	216.239.36.10
ns4.google.com.		172800	IN	AAAA	2001:4860:4802:38::a
ns4.google.com.		172800	IN	A	216.239.38.10
```

Scottmitch replied:
fine as is ... just for more context we can also account for the ADDITIONAL count up front too ... but accounting up front will only ever be an estimate and may require resize (tradeoff is *maybe* resize or over allocation vs *always* additional iteration).

normanmaurer commented:
@Scottmitch PTAL again

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DefaultAuthoritativeDnsServerCache.java:
```diff
@@ -0,0 +1,127 @@
+
+        @Override
+        protected boolean equals(InetSocketAddress entry, InetSocketAddress otherEntry) {
+            return entry.getHostName().equalsIgnoreCase(otherEntry.getHostName());
```
This should never happen as we will only try to add if the hostname is resolved:

https://github.com/netty/netty/pull/8177/files#diff-7ab1cd3db010ccc37c01987e522675caR104

That said let me use getHostString when on java7+

Scottmitch replied:
IIUC we add unresolved entries and resolved entries to the `AuthoritativeDnsServerCache`. The reverse lookup occurs if the address is created via an IP, so seems like this would actually be an issue for the resolved cases.

normanmaurer replied:
we should only add entries which have a hostname

Scottmitch replied:
got it ... we always have/use the name server name (and not just the IP) 👍

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/Cache.java:
```diff
@@ -0,0 +1,290 @@
+                    }
+                }
+            } else {
+                set(Collections.singletonList(e));
```
done

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -829,50 +968,176 @@ void add(DnsRecord r) {
+                        for (int i = 0; i < entries.size(); i++) {
+                            InetAddress address = entries.get(i).address();
+                            if (address != null) {
+                                if (serverName.address != null) {
```
I added a unit test to prove all is working as expected (which is the same as what you expected as correct behaviour).

Scottmitch replied:
Oh, I see. Something special happens on the first iteration (`serverName.address` is set), then all other iterations follow a different path. Consider making this this more explicit to clarify the control flow (and avoid extra conditionals on subsequent iterations):

```diff
diff --git a/resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java b/resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java
index a83762c8fd..b1666029d7 100644
--- a/resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java
+++ b/resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java
@@ -1033,18 +1033,29 @@ abstract class DnsResolveContext<T> {
                         for (int i = 0; i < entries.size(); i++) {
                             InetAddress address = entries.get(i).address();
                             if (address != null) {
-                                if (serverName.address != null) {
-                                    AuthoritativeNameServer server = new AuthoritativeNameServer(serverName);
-                                    server.next = serverName.next;
-                                    serverName.next = server;
-                                    serverName = server;
-
-                                    resolvedCount++;
-                                }
                                 // Use a TTL of Long.MIN_VALUE as we don't want to cache the actual address later on as
                                 // we have no idea about how long the TTL should be. When we encounter Long.MIN_VALUE
                                 // later on we will cache the unresolved address.
                                 serverName.update(parent.newRedirectServerAddress(address), Long.MIN_VALUE);
+                                ++i;
+                                for (; i < entries.size(); i++) {
+                                    address = entries.get(i).address();
+                                    // TODO: can address be null at this point?
+                                    if (address != null) {
+                                        AuthoritativeNameServer server = new AuthoritativeNameServer(serverName);
+                                        server.next = serverName.next;
+                                        serverName.next = server;
+                                        serverName = server;
+
+                                        resolvedCount++;
+
+                                        // Use a TTL of Long.MIN_VALUE as we don't want to cache the actual address later on as
+                                        // we have no idea about how long the TTL should be. When we encounter Long.MIN_VALUE
+                                        // later on we will cache the unresolved address.
+                                        serverName.update(parent.newRedirectServerAddress(address), Long.MIN_VALUE);
+                                    }
+                                }
+                                break;
                             }
                         }
                     }

```

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -829,50 +968,176 @@ void add(DnsRecord r) {
+                        // the unresolved address for now. We will then query the cache again the next time we try to
+                        // use it for a query which will do the correct thing in terms of respecting the TTL of the
+                        // A / AAAA record.
+                        InetSocketAddress addressToCache = server.ttl == Long.MIN_VALUE ?
```
I think I would prefer to do what we do for now... If we find out its a problem we can relax it in a follow up.

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -829,50 +968,176 @@ void add(DnsRecord r) {
+         * Creates a new {@link List} which holds the {@link InetSocketAddress}es.
+         */
+        List<InetSocketAddress> cacheAndReturnAsList(AuthoritativeDnsServerCache cache, EventLoop loop) {
+            List<InetSocketAddress> addressList = new ArrayList<InetSocketAddress>(resolvedCount);
```
@Scottmitch `resolvedCount` should be at least as large as we need or even larger as we update it also when we add entries as part of resolving these from the `DnsCache`. Can you tell me why you think this is not the case ?

Scottmitch replied:
`handleWithoutAdditionals` will set `serverName.address` to an unresolved `InetSocketAddress` if no corresponding entries can be found in the cache without incrementing `resolvedCount`. IIUC the intention is to include elements which otherwise cannot be resolved (via additionals or cache) to avoid intermediate lookups in case of TTL differences. In this sense the `resolvedCount` seems to under count, and the "resolved" terminology seems to not account for everything we are interested in.

normanmaurer replied:
ah sorry you are right... just renamed to servers and ensure we always update when we add a server address.

normanmaurer commented:
@Scottmitch ready for another review + added some replies to your questions. I think we are close (hopefully) :)

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -829,50 +968,175 @@ void add(DnsRecord r) {
+                            InetAddress address = entries.get(i).address();
+                            if (address != null) {
+                                if (serverName.address != null) {
+                                    AuthoritativeNameServer server = new AuthoritativeNameServer(serverName);
```
IIUC with the "unrolled loop" this condition `serverName.address != null` can no longer occur? if so this code block can be removed.

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -829,50 +968,175 @@ void add(DnsRecord r) {
+                                for (; i < entries.size(); i++) {
+                                    address = entries.get(i).address();
+                                    // If the first address was not null the rest should not be null as well but
+                                    // let us just check.
```
+1 for the comment ... would this be a candidate for an `assert` or do you think it is legit?

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -453,34 +497,85 @@ private boolean handleRedirect(
-                    final AuthoritativeNameServer authoritativeNameServer = serverNames.remove(recordName);
+                    // We may have multiple ADDITIONAL entries for the same nameserver name. For example one AAAA and
+                    // one A record.
+                    serverNames.handleWithAdditional(parent, r, authoritativeDnsServerCache());
```
cache `authoritativeDnsServerCache()` in a local variable before loop?

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -829,50 +967,175 @@ void add(DnsRecord r) {
+
+        private static void cache(AuthoritativeNameServer server, AuthoritativeDnsServerCache cache, EventLoop loop) {
+            // Cache NS record if not for a root server as we should never cache for root servers.
+            if (!server.isRootServer() && server.address != null) {
```
`&& server.address != null` -> IIUC this is always true with the current code ... consider removing.

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -829,50 +967,175 @@ void add(DnsRecord r) {
+                // the unresolved address for now. We will then query the cache again the next time we try to
+                // use it for a query which will do the correct thing in terms of respecting the TTL of the
+                // A / AAAA record.
+                InetSocketAddress addressToCache = server.ttl == Long.MIN_VALUE ?
```
`server.ttl == Long.MIN_VALUE` -> is this still necessary? Can't we just use `server.address`?

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -829,50 +967,176 @@ void add(DnsRecord r) {
+                if (serverName.address == null) {
+                    // These will be resolved on the fly if needed.
+                    cacheUnresolved(serverName, authoritativeCache, parent.executor());
+                    servers++;
```
the `AuthoritativeNameServer ` isn't being modified here ... remove this increment?

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -829,50 +968,167 @@ void add(DnsRecord r) {
-            return count;
+        private static void cacheUnresolved(
+                AuthoritativeNameServer server, AuthoritativeDnsServerCache authoritativeCache, EventLoop loop) {
+            // As we have no idea about how long the TTL should be we will just cache the unresolved
```
remove this comment? TTL is not used anywhere in this method and this appears to be left over from previous refactoring.

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java:
```diff
@@ -785,7 +923,8 @@ private boolean query(String hostname, DnsRecordType type, DnsServerAddressStrea
         private AuthoritativeNameServer head;
-        private int count;
+
+        private int servers;
```
`servers` -> `nameServerCount`?
